File: src/chipschallenge/blockreactions/YellowLockTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Inventory.Key;

public class YellowLockTo extends NoSlipReaction {

    private YellowLockTo() {
    }
    private static YellowLockTo mInstance = null;

    public static synchronized YellowLockTo getInstance() {
        if (mInstance == null) {
            mInstance = new YellowLockTo();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) {
        useKey(Key.YELLOW);
        standing.destroy();
    }

    public boolean canMove(Block moving, Block standing) {
        return moving.isChip() && hasKey(Key.YELLOW);
    }
}

File: ChipsChallenge.Shared/Blockreactions/YellowLockTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Key = Inventory.Key;

    public class YellowLockTo : NoSlipReaction
    {
        private static YellowLockTo instance;

        private YellowLockTo()
        {
        }

        public static YellowLockTo Instance
        {
            get
            {
                lock (typeof(YellowLockTo))
                {
                    return instance ?? (instance = new YellowLockTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            UseKey(Key.YELLOW);
            standing.Destroy();
        }

        public override bool canMove(Block moving, Block standing)
        {
            return moving.Chip && HasKey(Key.YELLOW);
        }
    }
}